File: src/rust/math/pow.rs
pub fn pow (x: i32, n:i32) -> i32 {
  // initial condition
  if n == 1 {
    return x
  }

  // even number
  if (n % 2) == 0 {
    return pow(x * x, n / 2)
  }

  // odd number
  return x * pow(x * x, (n -1) / 2)
}

File: src/haskell/math/pow.hs
pow :: Int -> Int -> Int
pow x n
  -- initial condition
  | n == 1 = x

  -- even number
  | (n `rem` 2) == 0 = pow (x * x) (n `div` 2)

  -- odd number
  | otherwise = x * pow (x * x) ((n - 1) `div` 2)
